<a href="https://colab.research.google.com/github/Amyerm/ClassFiles/blob/main/Actividad_Calificada_07_Datos_Duplicados_271756.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | [completar] |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Ari Daniel Mendoza Enrriquez |
| **Matrícula** | 271756 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [27]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [28]:
# Comparar todas las columnas para identificar filas exactamente iguales.
duplicados_primero = hoteles.duplicated(keep='first').sum()
duplicados_ultimo = hoteles.duplicated(keep='last').sum()
duplicados_todos = hoteles.duplicated(keep=False).sum()

print('Conservar primera: ', duplicados_primero, 'filas sobrantes')
print('Conservar última: ', duplicados_ultimo, 'filas sobrantes')
print('Marcar todas:     ', duplicados_todos, 'filas de grupos repetidos')

Conservar primera:  31994 filas sobrantes
Conservar última:  31994 filas sobrantes
Marcar todas:      40165 filas de grupos repetidos


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [29]:
# Las filas sobrantes son las que quitaría si conservo una de cada grupo.
porcentaje_sobrantes = duplicados_primero / len(hoteles) * 100
# Cada grupo repetido aporta una fila conservada además de las sobrantes.
grupos_repetidos = duplicados_todos - duplicados_primero

print(f'Filas sobrantes: {porcentaje_sobrantes:.2f}%')
print('Grupos repetidos:', grupos_repetidos)

Filas sobrantes: 26.80%
Grupos repetidos: 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [30]:
# Marcar duplicados antes de agrupar, para comparar las 32 columnas originales.
revision_hoteles = hoteles.copy()
revision_hoteles['es_duplicado'] = hoteles.duplicated(keep='first')

# La media de los booleanos da la proporción dentro de cada hotel.
duplicados_por_hotel = revision_hoteles.groupby('hotel')['es_duplicado'].agg(['size', 'sum', 'mean'])
duplicados_por_hotel.columns = ['reservas', 'filas_sobrantes', 'proporcion']
duplicados_por_hotel['porcentaje'] = duplicados_por_hotel['proporcion'] * 100
print(duplicados_por_hotel[['reservas', 'filas_sobrantes', 'porcentaje']].to_string())

              reservas  filas_sobrantes  porcentaje
hotel                                              
City Hotel       79330            25902   32.650952
Resort Hotel     40060             6092   15.207189


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

Se encontraron 31,994 filas sobrantes de 119,390 (26.80 %) y hay 40,165 filas involucradas en 8,171 grupos repetidos. El porcentaje es considerable para analizarlo, pero no basta para concluir que todas son errores. No afecta igual a los hoteles: City Hotel tiene 25,902 de 79,330 filas sobrantes (32.65 %) y Resort Hotel 6,092 de 40,060 (15.21 %).

---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [31]:
# keep=False incluye también la primera fila de cada grupo repetido.
filas_repetidas = hoteles[hoteles.duplicated(keep=False)]
columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                  'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                  'deposit_type', 'adr', 'reservation_status']
# Ordenar por hotel, llegada y características de la reserva para juntar casos parecidos.
orden = ['hotel', 'arrival_date_year', 'arrival_date_month',
         'arrival_date_day_of_month', 'lead_time', 'adr', 'reservation_status']
print(filas_repetidas.sort_values(orden)[columnas_vista].head(12).to_string(index=False))

     hotel  lead_time  arrival_date_year arrival_date_month  arrival_date_day_of_month  adults country market_segment deposit_type  adr reservation_status
City Hotel          2               2015             August                          1       2     ITA  Offline TA/TO   No Deposit 94.5           Canceled
City Hotel          2               2015             August                          1       2     ITA  Offline TA/TO   No Deposit 94.5           Canceled
City Hotel          2               2015             August                          1       2     ITA  Offline TA/TO   No Deposit 94.5           Canceled
City Hotel        117               2015             August                          1       1     ITA  Offline TA/TO   No Deposit 72.5          Check-Out
City Hotel        117               2015             August                          1       1     ITA  Offline TA/TO   No Deposit 72.5          Check-Out
City Hotel        117               2015             August           

### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [32]:
# dropna=False permite incluir también combinaciones con agent o company faltantes.
tamanos_grupo = hoteles.value_counts(dropna=False)
grupos_con_copias = tamanos_grupo[tamanos_grupo > 1]

# Contar cuantos grupos tienen cada cantidad de copias.
frecuencia_tamanos = grupos_con_copias.value_counts()
print('Los 10 tamaños de grupo más frecuentes:')
print(frecuencia_tamanos.head(10).to_string())
print('\nCopias del grupo más grande:', grupos_con_copias.max())
print('Valores de todas sus columnas:')
print(grupos_con_copias.head(1).to_string())

Los 10 tamaños de grupo más frecuentes:
count
2     5034
3      969
4      497
5      275
6      171
10     123
7      115
8      103
20      97
9       93

Copias del grupo más grande: 180
Valores de todas sus columnas:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                        2                     2       0.0       0       BB    PRT      Gro

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [33]:
# Comparar porcentajes del dataset completo contra los grupos repetidos.
segmentos = pd.concat([
    hoteles['market_segment'].value_counts(normalize=True).rename('total'),
    filas_repetidas['market_segment'].value_counts(normalize=True).rename('repetidas')
], axis=1).fillna(0) * 100
depositos = pd.concat([
    hoteles['deposit_type'].value_counts(normalize=True).rename('total'),
    filas_repetidas['deposit_type'].value_counts(normalize=True).rename('repetidas')
], axis=1).fillna(0) * 100

print('Segmento de mercado (%):\n', segmentos.round(2).to_string())
print('\nTipo de depósito (%):\n', depositos.round(2).to_string())
print(f'\nCancelación total: {hoteles["is_canceled"].mean() * 100:.2f}%')
print(f'Cancelación en filas repetidas: {filas_repetidas["is_canceled"].mean() * 100:.2f}%')

Segmento de mercado (%):
                 total  repetidas
market_segment                  
Online TA       47.30      20.59
Offline TA/TO   20.29      30.40
Groups          16.59      41.61
Direct          10.56       3.50
Corporate        4.44       3.67
Complementary    0.62       0.19
Aviation         0.20       0.05
Undefined        0.00       0.00

Tipo de depósito (%):
               total  repetidas
deposit_type                  
No Deposit    87.65      64.17
Non Refund    12.22      35.64
Refundable     0.14       0.19

Cancelación total: 37.04%
Cancelación en filas repetidas: 58.39%


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

a) No se puede asegurar que las 31,994 filas sobrantes sean errores porque falta un ID de reserva. Hay 8,171 grupos repetidos y el mayor contiene 180 filas idénticas que son: City Hotel, segmento Groups, depósito Non Refund y estado Canceled. Una reserva de grupo puede generar varias habitaciones con características iguales y también puede haber cargas repetidas. En las filas involucradas, Groups representa 41.61 % frente a 16.59 % del total, Non Refund 35.64 % frente a 12.22 %, y la cancelación es 58.39 % frente a 37.04 %. Esto hace probable que parte de las coincidencias correspondan a reservas distintas o a un proceso común de cancelación y la limpieza exacta sirve para medir el efecto, pero no confirma el origen de cada repetición.

b) duplicated() considera iguales dos valores faltantes en una misma columna al comparar filas. Por eso puede marcar como repetidas filas que tienen NaN tanto en agent como en company, siempre que coincidan las demás columnas. No demuestra que tengan la misma agencia o empresa, falta esa información. En value_counts() usé dropna=False para no perder esos grupos del conteo.

---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [34]:
# Quitar solo copias exactas y renumerar el índice para la comparación solicitada.
hoteles_limpio = hoteles.drop_duplicates(keep='first', ignore_index=True)
print('Filas originales:', len(hoteles))
print('Filas después:', len(hoteles_limpio))
print('Filas eliminadas:', len(hoteles) - len(hoteles_limpio))

Filas originales: 119390
Filas después: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [35]:
# Calcular medias por hotel y el número de filas con los dos conjuntos de datos.
resumen_original = hoteles.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()
resumen_original['reservas'] = hoteles.groupby('hotel').size().values
resumen_limpio = hoteles_limpio.groupby('hotel')[['is_canceled', 'adr']].mean().reset_index()
resumen_limpio['reservas'] = hoteles_limpio.groupby('hotel').size().values

# Poner nombres fáciles de interpretar en la tabla final.
resumen_original = resumen_original.rename(columns={'is_canceled': 'tasa_cancelacion', 'adr': 'tarifa_promedio'})
resumen_limpio = resumen_limpio.rename(columns={'is_canceled': 'tasa_cancelacion', 'adr': 'tarifa_promedio'})
print('Original:\n', resumen_original.to_string(index=False))
print('\nSin duplicados:\n', resumen_limpio.to_string(index=False))

Original:
        hotel  tasa_cancelacion  tarifa_promedio  reservas
  City Hotel          0.417270       105.304465     79330
Resort Hotel          0.277634        94.952930     40060

Sin duplicados:
        hotel  tasa_cancelacion  tarifa_promedio  reservas
  City Hotel          0.300386       110.985944     53428
Resort Hotel          0.234809        99.025346     33968


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [36]:
# Etiquetar cada versión antes de unirlas y ordenar ambos hoteles.
resumen_original['version'] = 'original'
resumen_limpio['version'] = 'sin duplicados'
comparacion = pd.concat([resumen_original, resumen_limpio], ignore_index=True)
comparacion = comparacion.sort_values(['hotel', 'version']).reset_index(drop=True)
print(comparacion.to_string(index=False))

       hotel  tasa_cancelacion  tarifa_promedio  reservas        version
  City Hotel          0.417270       105.304465     79330       original
  City Hotel          0.300386       110.985944     53428 sin duplicados
Resort Hotel          0.277634        94.952930     40060       original
Resort Hotel          0.234809        99.025346     33968 sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

a) En volumen la cantidad de reservas cambia más: City Hotel pasa de 79,330 a 53,428 (25,902 menos) y Resort Hotel de 40,060 a 33,968 (6,092 menos). Entre las tasas, el cambio mayor es la cancelación de City Hotel: de 41.73 % a 30.04 %, una baja de 11.69 puntos porcentuales, Resort baja de 27.76 % a 23.48 % (4.28 puntos). La tarifa media sube de 105.30 a 110.99 en City Hotel y de 94.95 a 99.03 en Resort. Los repetidos tienen más cancelaciones (58.39 %) que el conjunto completo (37.04 %), por eso su eliminación baja la tasa. Esto describe el efecto del criterio elegido, no prueba que la tasa limpia sea la real.

b) Si hay copias de un mismo caso, el modelo da más peso a esos patrones, especialmente a cancelaciones de ciertos segmentos. Si una copia queda en entrenamiento y otra en prueba, el modelo puede reconocerla y dar una evaluación demasiado optimista. Antes de separar los datos tendría que revisar la identidad de las reservas y mantener juntas las que representen el mismo caso, también evitaría borrar reservas legítimas solo por ser iguales.

---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [37]:
# Revisar duplicados y que el índice nuevo sea consecutivo desde cero.
print('Duplicados exactos:', hoteles_limpio.duplicated().sum())
print('Índices repetidos:', hoteles_limpio.index.duplicated().sum())
print('Primer índice:', hoteles_limpio.index[0])
print('Último índice:', hoteles_limpio.index[-1])
print('Último índice esperado:', len(hoteles_limpio) - 1)
print('Índice consecutivo:', hoteles_limpio.index.equals(pd.RangeIndex(len(hoteles_limpio))))

Duplicados exactos: 0
Índices repetidos: 0
Primer índice: 0
Último índice: 87395
Último índice esperado: 87395
Índice consecutivo: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [38]:
# Comparar todas las columnas salvo la fecha del último cambio de estado.
columnas_sin_fecha = hoteles_limpio.columns.drop('reservation_status_date')
parciales_sobrantes = hoteles_limpio.duplicated(subset=columnas_sin_fecha).sum()
mascara_parciales = hoteles_limpio.duplicated(subset=columnas_sin_fecha, keep=False)
parciales = hoteles_limpio[mascara_parciales]
print('Filas sobrantes sin comparar la fecha:', parciales_sobrantes)
print('Filas involucradas:', len(parciales))

# Mostrar ambas copias para poder contrastar sus fechas de estado.
columnas_parciales = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                      'arrival_date_day_of_month', 'adr', 'reservation_status',
                      'reservation_status_date']
orden_parcial = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                 'arrival_date_day_of_month', 'adr', 'reservation_status_date']
print('\nEjemplos:')
print(parciales.sort_values(orden_parcial)[columnas_parciales].head(12).to_string(index=False))
print('\nEstado de las filas involucradas:')
print(parciales['reservation_status'].value_counts().to_string())

Filas sobrantes sin comparar la fecha: 245
Filas involucradas: 471

Ejemplos:
     hotel  lead_time  arrival_date_year arrival_date_month  arrival_date_day_of_month    adr reservation_status reservation_status_date
City Hotel          9               2015          September                         10  95.00           Canceled              2015-09-01
City Hotel          9               2015          September                         10  95.00           Canceled              2015-09-03
City Hotel         12               2017               July                         15 139.00           Canceled              2017-07-07
City Hotel         12               2017               July                         15 139.00           Canceled              2017-07-10
City Hotel         14               2015            October                         15 100.00           Canceled              2015-10-07
City Hotel         14               2015            October                         15 100.00       

**Respuesta 4.2:**

Al ignorar reservation_status_date aparecen 245 filas sobrantes y 471 filas involucradas, todas tienen estado Canceled. Pueden ser reservas distintas con los mismos atributos cuya cancelación se registró en fechas diferentes, o varias versiones de una misma reserva. La fecha distinta es información útil y no hay un ID que permita distinguir ambos casos. No las eliminaría automáticamente, revisaría un identificador de reserva o el historial de cambios antes de decidir.

### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

Comparé las 32 columnas de cada fila para detectar coincidencias exactas. La fecha reservation_status_date también forma parte de la comparación. Usé keep='first' e ignore_index=True. En las 32 columnas, las copias exactas tienen los mismos valores, así que conservar la primera o la última no cambia las métricas calculadas.
De 119,390 filas pasé a 87,396; eliminé 31,994 (26.80 %). Verifiqué cero duplicados exactos y un índice consecutivo de 0 a 87,395. Aunque el ejercicio considera una fila idéntica como duplicado para medir su impacto, al no existir un identificador de reserva, varias habitaciones o reservas diferentes podrían compartir los 32 valores, sobre todo en el segmento Groups. Los NaN coincidentes tampoco identifican clientes o agencias. Conservaría el original y pediría datos adicionales antes de aplicar esta eliminación a decisiones reales.

---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

No eliminaría automáticamente las copias si cada fila correspondiera a una habitación de una reserva de grupo. El grupo más grande tiene 180 filas iguales y el segmento Groups concentra 41.61 % de las filas repetidas, podrían ser transacciones válidas.
Pediría un ID único de reserva y, si una reserva puede incluir varias habitaciones, un ID de habitación o línea de reserva. También ayudarían la fecha y hora de creación y un historial de cambios, así podría distinguir una carga repetida de habitaciones distintas y entender por qué aparecen 245 coincidencias adicionales al ignorar la fecha de estado.
duplicated().sum() == 0 solo dice que no hay filas iguales en todas las columnas comparadas. Después de la limpieza exacta, aún quedan 245 filas sobrantes que coinciden en todo excepto reservation_status_date. También puede haber diferencias de formato que escondan coincidencias.

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.